### PR TITLE
hammer: rgw: TempURL in radosgw behaves now like its Swift's counterpart.

### PR DESCRIPTION
**Pull request by Werkov**
http://tracker.ceph.com/issues/18317

### CODE REVIEW

**Comment by smithfarm** [comment]
**original description**

Signed-off-by: Radoslaw Zarzynski <rzarzynski@mirantis.com>
(cherry picked from commit a1ea7180969dcbaf88dfcc1bfafb17e82639df64)

Conflicts:
	src/rgw/rgw_swift.cc (replace C++11 constructs, drop
	temp_url_make_content_disp() which has no use with hammer's
	req_state)

**Comment by smithfarm** [comment]
@Werkov Looks good, thanks. Will put it through an rgw suite.

**Comment by smithfarm** [comment]
@yehudasa @oritwas @mattbenjamin This passed a RGW suite at http://tracker.ceph.com/issues/17151#note-35

Do you think it's good to merge?

File: src/rgw/rgw_swift.cc [hunk 1]
```diff
@@ -589,32 +608,55 @@ int authenticate_temp_url(RGWRados *store, req_state *s)
...
+   * and old versions of RadosGW. The second item will have the prefix
+   * of Swift API entry point removed. */
+  const size_t pos = g_conf->rgw_swift_url_prefix.find_last_not_of('/') + 1;
+  vector<string> allowed_paths;
+  allowed_paths.push_back(s->info.request_uri);
+  allowed_paths.push_back(s->info.request_uri.substr(pos + 1));
 
-  dout(20) << "temp url signature (plain text): " << str << dendl;
+  vector<string> allowed_methods;
+  allowed_methods.push_back(s->info.method);
```
Comment (oritwas) [inline src/rgw/rgw_swift.cc]: doesn't this need be in the else?

**Comment by smithfarm** [comment]
@Werkov I think @oritwas is referring to this:

master:

```
+  if (strcmp("HEAD", s->info.method) == 0) {
+    /* HEAD requests are specially handled. */
+    allowed_methods = { s->info.method, "PUT", "GET" };
+  } else {
+    allowed_methods = { s->info.method };
+  }
```

backport:

```
+  if (strcmp("HEAD", s->info.method) == 0) {
+    /* HEAD requests are specially handled. */
+    allowed_methods.push_back("PUT");
+    allowed_methods.push_back("GET");
+  }
```

This difference is not explained in the conflict resolution?

**Comment by Werkov** [comment]
`s->info.method` is included in both branches (originally), so I pushed it to the vector unconditionally and got rid of the else branch at all. (It's because this Ceph version doesn't use C++11 initializer list ctors.) Do you need more verbose conflict resolution message (*"replace C++11 constructs"*).

**Comment by oritwas** [comment]
@Werkov , could you add a comment in the commit

**Comment by Werkov** [comment]
Commit amended.

**Comment by smithfarm** [comment]
This just passed another rgw suite, see: http://tracker.ceph.com/issues/17151#note-39

**Pull request merged**